### PR TITLE
High dimension sparse matrix

### PR DESCRIPTION
**Pull request by sadrasabouri**
#### Reference Issues/PRs
According to #185 from @kleysonr
#### What does this implement/fix? Explain your changes.
`sparse_table_print` function added to `pycm_output.py` using this function in 

- `print_matrix`
- `print_normalized_matrix`
- `save_stat`
functions with `sparse = True` flag you can print high dimension Confusion Matrix more suitable with ignoring rows and columns fully filled with 0s.

### CODE REVIEW

File: pycm/pycm_obj.py [hunk 1]
```diff
@@ -221,18 +234,32 @@ def save_stat(
...
+                normalized_matrix = "Normalized Matrix : \n\n" + \
+                                    sparse_table_print(self.classes,
+                                                self.normalized_table) + "\n\n"
+                one_vs_all = "\nOne-Vs-All : \n\n"
+                for c in self.classes:
+                    one_vs_all += str(c) + "-Vs-All : \n\n"
+                    [classes, table] = one_vs_all_func(self.classes, self.table,
+                                                    self.TP, self.TN, self.FP,
+                                                    self.FN, c)
+                    one_vs_all += sparse_table_print(classes, table) + "\n\n"
```
Comment (sepandhaghighi) [inline pycm/pycm_obj.py]: Why ? ?
One vs All mode doesn't need sparse printing.

**Comment by sepandhaghighi** [comment]
Also, add tests for `save_stat` and `print_normalized_matrix` methods

File: pycm/pycm_param.py [hunk 2]
```diff
@@ -40,6 +40,8 @@
 COMPARE_RESULT_WARNING = "Confusion matrices are too close and the best one can not be recognized."
 
 CLASS_NUMBER_WARNING = "The confusion matrix is a high dimension matrix and won't be demonstrated properly.\n" \
+                       "If confusion matrix has too many zeros (sparse matrix) you can True `sparse` flag in printing functions"\
+                       " then you'll have better demonstration.\n"\
```
Comment (sepandhaghighi) [inline pycm/pycm_param.py]: Modify this warning message.

File: pycm/pycm_output.py [hunk 3]
```diff
@@ -323,6 +323,53 @@ def table_print(classes, table):
...
+        row_sum = 0
+        col_sum = 0
+        for y in classes:
+            row_sum += table[x][y]
+            col_sum += table[y][x]
+        if row_sum == 0:
+            del sparse_table[x]
+            actual_classes.remove(x)
+        if col_sum == 0:
+            for row in classes:
```
[the remark this answers is not in the document]
Comment (sadrasabouri) [reply pycm/pycm_output.py]: Oh right it should be `actual_classes`

**Comment by sepandhaghighi** [comment]
About trailing white spaces issues in codacy : 
It's not a problem, but please remove trailing white spaces manually from test files.

Thanks

File: Test/output_test.py [hunk 4]
```diff
@@ -270,6 +270,72 @@
...
+1         0.5       0.5       0.0       0.0       
+<BLANKLINE>
+2         0.0       1.0       0.0       0.0       
+<BLANKLINE>
+4         0.0       0.0       0.0       1.0       
+<BLANKLINE>
+5         0.0       0.0       1.0       0.0       
+<BLANKLINE>
+<BLANKLINE>
+>>> save_stat_data = cm.save_stat("test")
```
[the remark this answers is not in the document]
Comment (sadrasabouri) [reply Test/output_test.py]: Excuse.

**Comment by sadrasabouri** [comment]
@sepandhaghighi 
I've duplicated the output sparse printing functions in order to test it's probable cache bugs in the recent commit.

File: pycm/pycm_util.py [hunk 5]
```diff
@@ -211,6 +242,37 @@ def normalized_table_calc(classes, table):
     return new_table
 
 
+def sparse_matrix_calc(classes, table):
```
Comment (sepandhaghighi) [inline pycm/pycm_util.py]: Duplication :

Duplicate of `sparse_table_calc` function.
`table` and `matrix`are one.

File: pycm/pycm_obj.py [hunk 6]
```diff
@@ -220,12 +243,25 @@ def save_stat(
...
-            normalized_matrix = "Normalized Matrix : \n\n" + \
-                                table_print(self.classes,
-                                            self.normalized_table) + "\n\n"
+            if sparse is True:
+                if self.sparse_matrix is None:
+                    self.sparse_matrix = sparse_matrix_calc(classes, table)
+                matrix = "Matrix : \n\n" + \
+                    sparse_table_print(self.sparse_matrix) + "\n\n"
+                if self.sparse_normalized_matrix is None:
+                    table = normalized_table_calc(classes, table)
```
Comment (sepandhaghighi) [inline pycm/pycm_obj.py]: Why recalculation?
We have `self.normalized_table`

File: pycm/pycm_param.py [hunk 7]
```diff
@@ -40,8 +40,8 @@
 COMPARE_RESULT_WARNING = "Confusion matrices are too close and the best one can not be recognized."
 
 CLASS_NUMBER_WARNING = "The confusion matrix is a high dimension matrix and won't be demonstrated properly.\n" \
-                       "The save_csv method can be used to save the confusion matrix in csv format and have a better" \
-                       " demonstration of it."
+                       "If confusion matrix has too many zeros (sparse matrix) you can True `sparse` flag in printing functions "\
```
Comment (sepandhaghighi) [inline pycm/pycm_param.py]: you can set `sparse` flag to True ...

Comment (sadrasabouri) [reply pycm/pycm_param.py]: Thanks

File: pycm/pycm_obj.py (hunk 6 above)
Comment (sadrasabouri) [reply pycm/pycm_obj.py]: I hadn't realized `self.normalized_table` as an attribute :smile:

**Pull request merged**